Auth token package and middleware refactoring

marefr opened the pull request:
Ref #15067 

This PR fixes:
* moving middleware/hooks away from auth package to middleware/api packages
* exposing public struct UserToken accessible from other packages (api/middleware)
* changing configuration settings for login and auth token - remove new section login and resuses existing security and auth sections
* now there's a rolling maximum lifetime since token rotated and a maximum lifetime since token created
  * login_maximum_inactive_lifetime_days = 7
  * login_maximum_lifetime_days = 30
* clean up job should be configured to run on certain schedule and delete all expired tokens
* documentation updated due to configuration settings
* fix debug log lines in auth package so the same order and naming are used

marefr committed:
auth package refactoring

moving middleware/hooks away from package
exposing public struct UserToken accessible from other packages
fix debug log lines so the same order and naming are used

marefr committed:
move auth token middleware/hooks to middleware package

fix/adds auth token middleware tests

marefr committed:
remove unused code

marefr committed:
refactor login/auth token configuration settings

remove login section and reuse existing sections security and auth

marefr committed:
document login, short-lived tokens and secure cookie configurations

marefr committed:
change configuration settings in auth package

marefr committed:
changes needed for api/middleware due to configuration settings

marefr committed:
auth token clean up job now runs on schedule and deletes all expired tokens

delete tokens having created_at <= LoginMaxLifetimeDays or
rotated_at <= LoginMaxInactiveLifetimeDays

marefr committed:
enhanced expiration logic for lookup token

tokens are not expired if created_at > now - LoginMaxLifetimeDays and
rotated_at > now - LoginMaxInactiveLifetimeDays

marefr committed:
middleware fix

bergquist commented on pkg/services/auth/auth.go:
```diff
@@ -0,0 +1,6 @@
+package auth
+
+type UserToken interface {
```
Another idea would be to put a function on the ReqConext for getting the token from the cookie. 

Moving the functions for writing/read auth cookies to the reqContext might be an ok idea.

marefr replied:
Decided to change to a struct

marefr committed:
fix spelling

bergquist commented on conf/defaults.ini:
```diff
@@ -260,6 +244,21 @@ external_manage_info =
 viewers_can_edit = false
 
 [auth]
+# Login cookie name
```
having these settings under `[auth]` makes sense :)

Perhaps we should rename the `login` prefix?

bergquist commented on conf/defaults.ini:
```diff
@@ -260,6 +244,21 @@ external_manage_info =
+# The lifetime (days) an authenticated user can be inactive before being required to login at next visit. Default is 7 days.
+login_maximum_inactive_lifetime_days = 7
+
+# The maximum lifetime (days) an authenticated user can be logged in since login time before being required to login. Default is 30 days.
```
have we decided on this?

Forcing people to login every X breaks tv screens.

marefr replied:
I'm going to create a separate issue for this

marefr replied:
#15287

bergquist commented on pkg/api/login.go:
```diff
@@ -176,7 +185,8 @@ func (hs *HTTPServer) trySetEncryptedCookie(ctx *m.ReqContext, cookieName string
 		HttpOnly: true,
 		Path:     setting.AppSubUrl + "/",
-		Secure:   hs.Cfg.SecurityHTTPSCookies,
+		Secure:   hs.Cfg.CookieSecure,
```
having a helper function on ReqContext to make sure we always those these settings might not be a bad idea. Query to forget.

marefr replied:
I have a WIP cookie store stashed that we can implement in separate PR.

bergquist commented on pkg/services/auth/authtoken/session_cleanup.go:
```diff
@@ -0,0 +1,54 @@
+		return nil
+	}
+
+	jobInterval := time.Duration(srv.Cfg.ExpiredTokensCleanupIntervalDays) * 24 * time.Hour
```
As you mentioned during our last convo this requires the application to be running for 7 days until we clean up old sessions. I think we can run this at startup (since it's not blocking the server) and then every hour. 

I dont see any reason why someone would configure how often this runs?

marefr replied:
Changed it so cleanup is run when Grafana starts. Change interval to 1 hour,

marefr committed:
change UserToken from interface to struct

marefr committed:
move UserToken and UserTokenService to models package

marefr committed:
move authtoken package into auth package

marefr committed:
fix

marefr committed:
run token cleanup job when grafana starts, then each hour

marefr committed:
devenv: update ha test and load test

better ha setup for many mysql connections
prometheus now scrapes mysql metrics in ha setup
ha setup provisions mysql dashboard
adds configurable virtual users for load test run script

marefr commented:
### Load test against ha setup
#### Setup
Grafana instances=3
Grafana max open connections=200
Grafana token rotation interval=2 minutes
MySQL max connections=1001
Grafana users (devices)=100

Load test iteration steps:
1. First iteration? Signing in
2. Not first iteration? Requests 1 annotation query and 20 tsdb queries in paralell
3. Sleeps for 3 seconds
4. Repeat from 2)

#### Load test results
```bash

          /\      |‾‾|  /‾‾/  /‾/
     /\  /  \     |  |_/  /  / /
    /  \/    \    |      |  /  ‾‾\
   /          \   |  |‾\  \ | (_) |
  / __________ \  |__|  \__\ \___/ .io

  execution: local
     output: -
     script: src/auth_token_test.js

    duration: 15m0s, iterations: -
         vus: 100,   max: 100

WARN[0805] Request Failed                                error="Post http://grafana.loc/api/tsdb/query: net/http: request canceled (Client.Timeout exceeded while awaiting headers)"
WARN[0805] Request Failed                                error="Post http://grafana.loc/api/tsdb/query: net/http: request canceled (Client.Timeout exceeded while awaiting headers)"
    done [==========================================================] 15m0s / 15m0s

    █ user auth token test

      █ user authenticates thru ui with username and password

        ✓ response status is 200
        ✓ response has cookie 'grafana_session' with 32 characters

      █ batch tsdb requests

        ✗ response status is 200
         ↳  99% — ✓ 300939 / ✗ 2

    checks.....................: 99.99% ✓ 300938 ✗ 2
    data_received..............: 3.8 GB 4.2 MB/s
    data_sent..................: 110 MB 123 kB/s
    group_duration.............: avg=931.75ms min=25.56ms med=571.92ms max=1m0s     p(90)=1.97s   p(95)=3.21s
    http_req_blocked...........: avg=13.46ms  min=899ns   med=2.92ms   max=1s       p(90)=28.51ms p(95)=52.22ms
    http_req_connecting........: avg=11.9ms   min=0s      med=2.34ms   max=833.02ms p(90)=25.6ms  p(95)=45.98ms
    http_req_duration..........: avg=462.86ms min=2.59ms  med=242.48ms max=16.58s   p(90)=1.14s   p(95)=1.51s
    http_req_receiving.........: avg=132.42µs min=0s      med=33.1µs   max=365.71ms p(90)=84.18µs p(95)=181.92µs
    http_req_sending...........: avg=3.18ms   min=7.73µs  med=139.29µs max=518.07ms p(90)=4.75ms  p(95)=13.84ms
    http_req_tls_handshaking...: avg=0s       min=0s      med=0s       max=0s       p(90)=0s      p(95)=0s
    http_req_waiting...........: avg=459.54ms min=0s      med=239.48ms max=16.57s   p(90)=1.14s   p(95)=1.5s
    http_reqs..................: 316109 351.2322/s
    iteration_duration.........: avg=5.94s    min=5.25µs  med=5.58s    max=1m5s     p(90)=6.98s   p(95)=8.23s
    iterations.................: 15055  16.727777/s
    vus........................: 100    min=100  max=100
    vus_max....................: 100    min=100  max=100

```

#### Grafana metrics
![grafana-metrics](https://user-images.githubusercontent.com/1668778/52375427-ea240680-2a5f-11e9-828c-f62f1ad7ecc9.png)

#### MySQL metrics
![mysql-metrics](https://user-images.githubusercontent.com/1668778/52375468-045de480-2a60-11e9-88c4-a04f538ae499.png)

bergquist commented on docs/sources/auth/overview.md:
```diff
@@ -64,8 +64,8 @@ login_maximum_lifetime_days = 30
-# How often should expired auth tokens be deleted from the database. The default is 7 days.
-expired_tokens_cleanup_interval_days = 7
+# How often should expired auth tokens be deleted from the database. The default is each hour.
+expired_tokens_cleanup_interval_hours = 1
```
can we remove this setting? Fewer settings are better admin UX.

bergquist commented:
Sweet looks good! I still want to remove the cleanup setting but otherwise, Lets Get This Merged!

bergquist committed:
make hourly cleanup the default behavior

marefr reviewed (commented):
Looks good. Need to remove expired_tokens_cleanup_interval_hours reference from auth/overview.md docs as well, but then should be ready for merge.

bergquist committed:
removes cleanup setting from docs

bergquist commented on pkg/services/auth/model.go:
```diff
@@ -23,3 +20,51 @@ type userAuthToken struct {
+	return &uat
+}
+
+func (uat *userAuthToken) fromUserToken(ut *models.UserToken) error {
```
I think this is noisy. Since we couldn't isolate this struct completely I think we should use `models.UserToken` in all places.

bergquist replied:
Not a blocker. So I'm fine merging this first and fixing it later.

marefr replied:
Somewhat agree that it's noisy, but still internal to the package I think it's okay. One of the nice things is that the database model is internal.

bergquist commented:
Push the green button! :)

lukeyeager commented:
Consider adding documentation about
```
[auth]
login_cookie_name = grafana_session
```
here: https://grafana.com/docs/grafana/latest/installation/configuration/

marefr commented:
@lukeyeager the documentation is already here https://grafana.com/docs/grafana/latest/auth/overview/#settings
